### ISSUE TITLE
Basic fact finding for Blackboard local file integration

### ISSUE THREAD

**Issue by jon-betts** [hypothesis/lms#1319]
In order to guide future development effort we should gather some basic details about each of the integrations before we start to ensure we don't go down any blind alleys.

### What is the integration method? JSON API? XML RPC?

 * JSON API

### Where are the docs?

* https://developer.blackboard.com/portal/displayApi
* https://github.com/blackboard/BBDN-REST-DEMO_Python - This doesn't look like nice code

### How do we authenticate?
   * Looks like there are OAuth2 end-points so I guess like that
   * The docs mention Three Legged Oauth (3LO) which is a new one on me
   * https://help.blackboard.com/Learn/Administrator/Hosting/System_Integration/Building_Blocks_and_REST_APIs/Three_Legged_OAuth
   * I think this might be a fancy way of saying the regular grant flow

### How do you get the list of files?

 * `GET /learn/api/public/v1/courses/{courseId}/contents/{contentId}/attachments`

### How do you get the file?

 * `GET /learn/api/public/v1/courses/{courseId}/contents/{contentId}/attachments/{attachmentId}/download`

### Do we need to proxy the file or can we return a direct URL?

  * Looks like we might get a direct file URL we can pass to the user
  * The JSON payload mentions `fileName` although this might just be the end part like 'file.pdf'
  * There is also a download end-point for a given attachment (but might pass the wrong content-header prompting download not embed)
  * You can also enable Cross-Origin support (not sure if this is what it's for)
  * https://help.blackboard.com/Learn/Administrator/Hosting/System_Integration/Cross_Origin_Resource_Sharing

### Does a particular assignment have one list of files or many?

 * One list per `courseId` + `contentId` which I'm guesing identifies an assignment
 *  `GET /learn/api/public/v1/courses/{courseId}/contents/{contentId}/attachments`

**Comment by jon-betts** [comment]
I did have an additional thought which I presume we already know based on how OAuth works in Canvas, but we need the teacher to login for us to retrieve the file, but does the student need to login to see it?

**Comment by robertknight** [comment]
> but does the student need to login to see it?

Yes. Anyone who views the assignment as either an instructor or student will need to authorize via OAuth to enable the lms app to fetch the file. Authorization only has to be completed once per Canvas instance AFAIK and only repeated if the credentials are no longer usable for any reason (they were revoked, they don't have the necessary permissions in some context etc.)

**Comment by seanh** [comment]
If it's possible to not proxy the Blackboard PDFs (if they don't have CORS restrictions etc) I think we'd have to add a new feature for that to py-proxy.

Currently (or once Via 3 a.k.a. py-proxy is enabled for LMS PDFs, anyway) the LMS app injects an `<iframe src="https://via3.hypothes.is/<UNPROXIED_PDF_URL>"`. Via 3 then:

1. Sends a `GET` request to decide whether `<UNPROXIED_PDF_URL>` is a PDF or not
2. Redirects the browser to a `via3.hypothes.is/pdf/<UNPROXIED_PDF_URL>` page that serves up the PDF.js PDF viewer app
3. Configures PDF.js to download the PDF file from `https://via3.hypothes.is/static/<UNPROXIED_PDF_URL>`. This is where the decision of whether to proxy the PDF or not gets made. The decision is made by Via 3, not by LMS. And Via 3 always proxies the PDF.

We've discussed before how, because the LMS app already knows it's a PDF, it could bypass (1) by injecting  a `<iframe src="https://via3.hypothes.is/pdf/<UNPROXIED_PDF_URL>"` directly (notice the additional `/pdf/`). But that's a separate issue.

For telling Via 3 not to proxy the PDF: Via 3 already includes support for a couple of different `via.*` query params that LMS can put in the iframe src URL. So maybe we can add support for a `via.do_not_proxy` param or something like that. Or we could use a separate Via 3 URL subpath "do not proxy".

But anyway, so I think we'd have to add support to Via 3 for optionally being told that it doesn't need to proxy the PDF. Then the LMS app could do whatever is necessary to trigger that option.

* * *

By the way: whenever step (1) is not being optimised out, so Via 3 is doing the extra `GET` request and downloading the PDF response's headers to decide whether it's a PDF or not, I guess Via 3 could also look at the CORS headers in the response and decide for itself whether the PDF needs to be proxied or not?

This may not apply to the LMS app if the LMS app is optimising step 1 away. But in other situations (or as long as we haven't yet got round to the LMS app skipping step 1 anyway) it might help

**Comment by seanh** [comment]
> I did have an additional thought which I presume we already know based on how OAuth works in Canvas, but we need the teacher to login for us to retrieve the file, but does the student need to login to see it?

The way this works in Canvas is that both the teacher and the student have been given access to the PDF file in Canvas. So when a student launches a Canvas file assignment we make Canvas API requests on behalf of the student (using a Canvas API access token for the student) in order to get a download URL for the PDF file. This does require each individual student to have granted us OAuth 2 access to their account, so each student will see the authorization popup window (once).

It is actually possible in Canvas for a teacher to create an assignment for a PDF file that the students do not have access to. Or to revoke student access to the file after creating the assignment. Or in fact to entirely delete the file from Canvas after creating the assignment. These sorts of things will result in errors when a student launches the assignment. See test assignments in hypoothesis.instructure.com's Test Course 1 such as _localhost (make devdata) students don't have access to Canvas file_ and _localhost (make devdata) Canvas file has been deleted_.

Note that it would not be correct for us to use the teacher's OAuth 2 access token to get the PDF file download URL when a student is launching the assignment. If the teacher has revoked student access to the file, for example, then they have revoked it, students should not still be able to access the file through Hypothesis. It would also cause problems when, during assignment launch, a new OAuth 2 authorization was required.

**Comment by dwhly** [comment]
Great to see this discussion.  I wondered last year whether in certain circumstances we could just iframe the sidebar over a PDF that would stream directly into the users browser w/o proxy.  Glad that it's actually a possibility.

**Comment by jon-betts** [comment]
Just a note to say that me and Sean discussed the proxying issue on a call, and it seems I've caused a bit of confusion.

In this ticket I am discussing whether LMS itself needs to proxy the file to make it accessible rather than Via. I am assuming Via proxying will always be necessary.

I was concerned about this because I could imagine a few situations where it might not be possible to get at the file directly:

 * The file is only accessible via the API
 * The file isn't available via HTTP (some mad other protocol in place)
 * The file is served with 'content-disposition' headers that trigger download, rather than viewing in browsers
 * Something else I haven't thought of yet, etc...

**Comment by jon-betts** [comment]
Also a quick note to say I think Sean has looked over this and deemed it "sane" if not correct.
